ambry-commons/src/main/java/com/github/ambry/compression/BaseCompression.java

```diff
@@ -306,4 +323,146 @@ public int decompress(byte[] compressedBuffer, int compressedBufferOffset, int c
+        // the compressed buffer contains a version that is not supported.  It's probably a code problem, not data problem.
+        // Modify this condition as new versions are added.
+        throw new CompressionException("compressedBuffer has unsupported version " + version);
+      }
+
+      // Assume the algorithm name is correct, so skip the name verification.
+      // Original size offset = version (1 byte) + name size (1 byte) + name (N byte) + data size (4 bytes);
+      int nameSizeInBytes = compressedBuffer.get();
+      int dataSizePosition = compressedBuffer.position() + nameSizeInBytes;
+      if (dataSizePosition + 4 >= compressedBuffer.limit()) {
+        throw new IllegalArgumentException("compressedBuffer too small and does not contain original data size.");
+      }
+      compressedBuffer.position(dataSizePosition);
+      return compressedBuffer.getInt();
+    } finally {
+      compressedBuffer.reset();
+    }
+  }
+
+  /**
+   * Compress the buffer specified in {@code sourceBuffer}.  The entire sourceBuffer will be read,
+   * so set the position and limit to the range of data to read in sourceBuffer before calling this method.
+   * After calling, sourceBuffer position will be advanced to the buffer limit, and the compressedBuffer position
+   * will be advanced to the end of the compressed binary.
+   * If compression failed, the sourceBuffer position will not be advanced.
+   *
+   * @param sourceBuffer The source/uncompressed data to compress.  It cannot be null or empty.
+   * @param compressedBuffer The compressed buffer where the compressed data will be written to.  Its size must be
+   *                       at least the size return from getCompressBufferSize() or it may fail due to buffer size.
+   * @return The actual compressed data size in bytes.
+   * @throws CompressionException Throws this exception if compression has internal failure.
+   */
+  @Override
+  public int compress(ByteBuffer sourceBuffer, ByteBuffer compressedBuffer) throws CompressionException {
+    Utils.checkNotNullOrEmpty(sourceBuffer, "sourceBuffer cannot be null or empty.");
+    Utils.checkNotNullOrEmpty(compressedBuffer, "compressedBuffer cannot be null or empty.");
+
+    int overheadSize = VERSION_AND_ORIGINAL_SIZE_SIZE + ALGORITHM_NAME_LENGTH_SIZE + getAlgorithmNameBinary().length;
+    if (compressedBuffer.remaining() < overheadSize) {
+      throw new IllegalArgumentException("compressedBuffer " + compressedBuffer.remaining() + " is too small.");
+    }
+
+    // Write version, algorithm name, and source data size.
+    int compressedBufferStartPosition = compressedBuffer.position();
+    int sourceBufferStartPosition = sourceBuffer.position();
+    try {
+      byte[] algorithmNameBinary = getAlgorithmNameBinary();
+      compressedBuffer.put((byte) 1);
+      compressedBuffer.put((byte) algorithmNameBinary.length);
+      compressedBuffer.put(algorithmNameBinary);
+      compressedBuffer.putInt(sourceBuffer.remaining());
```

Comment:
I vaguely remember that Java ByteBuffer's default byte order is big endian. But the method of compress (with byte array) is putting length in little endian order. This seems to be something concerning. Do we have a test for this?

Reply:
ByteBuffer default order is always big-endian order no matter what OS it is running in (OS independent).    Only calling order(newOrder) could alter that instance of ByteBuffer.  None of our code calls that.  You were right that the old implementation used little-endian by taking encoding in own hand.  To simplify the code, I decided to use big-endian which is ByteBuffer default and Java JVM default.